src/ModelContextProtocol/Server/AIFunctionMcpServerTool.cs

```diff
@@ -220,6 +220,45 @@ options.OpenWorld is not null ||
         return newOptions;
     }
 
+    /// <summary>
+    /// Filters a JsonElement containing a schema to only include allowed properties: "type", "properties", and "required".
+    /// </summary>
+    private static JsonElement FilterJsonSchema(JsonElement schema)
+    {
+        using var memoryStream = new MemoryStream();
+        using var writer = new Utf8JsonWriter(memoryStream);
```

Comment:
I think it would probably be better to use JsonObject to build up the JSON DOM and then convert that to a JsonElement using [JsonNode.Deserialize<JsonElement>()(https://learn.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializer.deserialize?view=net-9.0#system-text-json-jsonserializer-deserialize-1(system-text-json-nodes-jsonnode-system-text-json-serialization-metadata-jsontypeinfo((-0))))

https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/use-dom#create-a-jsonnode-dom-with-object-initializers-and-make-changes

@eiriktsarpalis is probably the best person to ask about the right approach here though.

Reply:
> Edit: Could we also use [AIFunctionFactoryOptions.JsonSchemaCreateOptions.TransformSchemaNode](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.ai.aijsonschemacreateoptions?view=net-9.0-pp) for the AIFunctions we create ourselves from Delegates and the like and avoid generating the intermediate schema in that case?

Yep, that's the best way to achieve that. Although per my remarks in https://github.com/modelcontextprotocol/csharp-sdk/pull/343#issuecomment-2824136953 this should most likely be left to individual implementers as I think the metaschema is not meant to be interpreted literally.